tensorflow/core/kernels/conv_ops_gpu_3.cu.cc

```diff
@@ -413,6 +429,275 @@ struct PadInput<GPUDevice, T, int, NDIMS> {
+                         TileLongSide == 512  && (TileShortSide OP 2)) || \
+      size_of_t == 4 && (TileLongSide == 32   && (TileShortSide OP 15) || \
+                         TileLongSide == 64   && (TileShortSide OP 15) || \
+                         TileLongSide == 128  && (TileShortSide OP 15) || \
+                         TileLongSide == 256  && (TileShortSide OP 10) || \
+                         TileLongSide == 512  && (TileShortSide OP 4)  || \
+                         TileLongSide == 1024 && (TileShortSide OP 2)) || \
+      size_of_t == 2 && (TileLongSide == 32   && (TileShortSide OP 15) || \
+                         TileLongSide == 64   && (TileShortSide OP 15) || \
+                         TileLongSide == 128  && (TileShortSide OP 15) || \
+                         TileLongSide == 256  && (TileShortSide OP 10) || \
+                         TileLongSide == 512  && (TileShortSide OP 4)  || \
+                         TileLongSide == 1024 && (TileShortSide OP 2)) || \
+      size_of_t == 1 && (TileLongSide == 32   && (TileShortSide OP 15) || \
+                         TileLongSide == 64   && (TileShortSide OP 15) || \
+                         TileLongSide == 128  && (TileShortSide OP 15) || \
+                         TileLongSide == 256  && (TileShortSide OP 10) || \
+                         TileLongSide == 512  && (TileShortSide OP 4)  || \
+                         TileLongSide == 1024 && (TileShortSide OP 2))
+
+constexpr bool TileSizePossibilityFrontierCheck(int TileLongSide,
+                                                int TileShortSide,
+                                                int size_of_t, int mode) {
+  // When mode is 0, this function checks whether the tile size combination lies
+  // on the tile size possibility frontier. When mode is 1, it checks whether
+  // the combination lies outside the frontier.
+  return mode == 0 ? TPF_CHECKER(==) : TPF_CHECKER(>);
+}
+
+#undef TPF_CHECKER
+
+// Recursive template function to search for the minimum tile size configuration
+// satisfying the requested tile side lengths.
+template <typename T, int TileLongSide, int TileShortSide, typename dummy = void>
+struct BatchNarrowMatrixTransposeDispatcher {
+  static void DoIt(const GPUDevice& d, int tile_size_i,
+                                           int tile_size_j,
+                                           int total_tiles_count,
+                                           const T* input,
+                                           const Dimension<3>& input_dims,
+                                           T* output) {
+
+    static_assert((TileLongSide & (TileLongSide - 1)) == 0,
+      "The length of the longer side of the tile is always a power of 2.");
+    bool request_satisfied = max(tile_size_i, tile_size_j) <= TileLongSide &&
+                             min(tile_size_i, tile_size_j) <= TileShortSide;
+
+    if (request_satisfied) {
+      const int NumThread = TileLongSide;
```

Comment:
Although this works, I think constexpr would be better than plain const, to emphasize the fact that we're going to use this as a template parameter so it must be knowable at compile-time.